lib/filterx/filterx-metrics.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2024 Axoflow
+ * Copyright (c) 2024 Attila Szakacs <attila.szakacs@axoflow.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx-metrics.h"
+#include "filterx-metrics-labels.h"
+#include "expr-literal.h"
+#include "object-string.h"
+#include "filterx-eval.h"
+#include "stats/stats-cluster-single.h"
+#include "stats/stats-registry.h"
+#include "metrics/metrics-tls-cache.h"
+#include "scratch-buffers.h"
+
+struct _FilterXMetrics
+{
+  struct
+  {
+    FilterXExpr *expr;
+    gchar *str;
+  } key;
+
+  FilterXMetricsLabels *labels;
+  gint level;
+
+  StatsCluster *const_cluster;
+};
+
+gboolean
+filterx_metrics_is_enabled(FilterXMetrics *self)
+{
+  return stats_check_level(self->level);
+}
+
+static const gchar *
+_format_sck_name(FilterXMetrics *self)
+{
+  if (self->key.str)
+    return self->key.str;
+
+  FilterXObject *key_obj = filterx_expr_eval_typed(self->key.expr);
+  if (!key_obj)
+    return NULL;
+
+  gsize len;
+  const gchar *name = filterx_string_get_value(key_obj, &len);
+  if (len == 0)
+    {
+      filterx_eval_push_error("failed to format metrics key: key must be a non-empty string", self->key.expr, key_obj);
+      goto exit;
+    }
+
+  GString *name_buffer = scratch_buffers_alloc();
+  g_string_append(name_buffer, name);
```

Comment:
I guess you were trying to avoid a g_strdup() here. in that case, it makes sense to use g_string_append_len(), otherwise you'd be doing an strlen() as a part of the copy, while the length is already available.

Reply:
Thanks, fixed and squashed to the original commit.